geomesa-features/geomesa-feature-common/src/main/scala/org/locationtech/geomesa/features/ScalaSimpleFeature.scala

```diff
@@ -146,21 +148,34 @@ object ScalaSimpleFeature {
     * @param readAttribute lazily read an attribute value, must already be converted into the appropriate types
     * @param readUserData lazily read the user data
     */
-  class LazyImmutableSimpleFeature(sft: SimpleFeatureType,
-                                   id: String,
-                                   readAttribute: Int => AnyRef,
-                                   readUserData: => java.util.Map[AnyRef, AnyRef])
-      extends AbstractImmutableSimpleFeature(sft, id) {
-
-    private val bits = AtomicBitSet(sft.getAttributeCount)
-    private val attributes = Array.ofDim[AnyRef](sft.getAttributeCount)
-
-    override lazy val getUserData: java.util.Map[AnyRef, AnyRef] =
-      Collections.unmodifiableMap(synchronized { readUserData })
+  class LazyImmutableSimpleFeature(
+      sft: SimpleFeatureType,
+      id: String,
+      private var readAttribute: Int => AnyRef,
+      private var readUserData: () => java.util.Map[AnyRef, AnyRef]
+    ) extends AbstractImmutableSimpleFeature(sft, id) {
+
+    private val bits = IntBitSet(sft.getAttributeCount)
+    private lazy val attributes = Array.ofDim[AnyRef](sft.getAttributeCount)
+    private var count = 0
+
+    override lazy val getUserData: java.util.Map[AnyRef, AnyRef] = {
+      val read = synchronized { readUserData() }
+      readUserData = null // dereference any backing resources
+      Collections.unmodifiableMap(read)
+    }
 
     override def getAttribute(index: Int): AnyRef = {
-      if (bits.add(index)) {
-        attributes(index) = synchronized { readAttribute(index) }
+      if (readAttribute != null) {
+        bits.synchronized {
+          if (bits.add(index)) {
+            attributes(index) = synchronized { readAttribute(index) }
```

Comment:
Do we need to worry about all the synchronized calls in here?

Reply:
do you mean the nested synchronized blocks, or just the synchronized blocks in general? I am fairly certain the nested blocks can't cause a deadlock, as the inner locks are not dependent on anything else. we need to synchronize on reading attributes/user data, as it uses the same single-threaded kryo reader under the hood. previously we were using an AtomicBitSet to avoid synchronizing on the bitset, but I think that wasn't entirely thread-safe in that it could potentially return null attributes while an attribute was being read. the synchronized blocks here are guarded by a null check, so once all the attributes have been read it won't synchronize on access anymore. I am fairly sure that is safe to do, since we are only setting `readAttribute` null inside a synchronized block, after all attributes are read.

Reply:
so, in short, yes we should worry about it but it's necessary and I think it's implemented correctly :)